sealtk/core/VideoController.cpp

```diff
@@ -44,62 +51,113 @@ VideoController::~VideoController()
 {
   QTE_D();
-  return d->videoSources;
+
+  auto out = QSet<VideoSource*>{};
+  out.reserve(d->videoSources.size());
+
+  for (auto const& item : qtEnumerate(d->videoSources))
+  {
+    out.insert(item.key());
+  }
+
+  return out;
 }
 
 // ----------------------------------------------------------------------------
-void VideoController::addVideoSource(VideoSource* videoSource)
+VideoDistributor* VideoController::distributor(VideoSource* videoSource) const
 {
   QTE_D();
+  return d->videoSources.value(videoSource);
+}
 
-  using Time = kwiver::vital::timestamp::time_t;
+// ----------------------------------------------------------------------------
+VideoDistributor* VideoController::addVideoSource(VideoSource* videoSource)
+{
+  QTE_D();
 
-  if (!d->videoSources.contains(videoSource))
+  if (auto* const existingDistributor = d->videoSources.value(videoSource))
   {
-    d->videoSources.insert(videoSource);
-    connect(this, &VideoController::timeSelected,
-            videoSource, &VideoSource::seekTime);
+    return existingDistributor;
+  }
 
-    if (d->videoSources.size() == 1)
-    {
-      auto const& frames = videoSource->frames();
-      if (!frames.empty())
-      {
-        seek(frames.begin().key());
-      }
-    }
-    else
-    {
-      videoSource->seekTime(d->time);
-    }
+  connect(videoSource, &VideoSource::framesChanged, this,
+          [this, d]{
+            d->timesDirty = true;
 
-    emit this->videoSourcesChanged();
+            if (!d->timeIsValid)
+            {
+              d->updateTimes();
+              if (!d->times.isEmpty())
+              {
+                this->seek(d->times.begin().key());
+              }
+            }
+
+            emit this->timesChanged();
+          });
+
+  connect(videoSource, &QObject::destroyed, this,
+          [this, videoSource]{
+            this->removeVideoSource(videoSource);
+          });
+
+  auto fetch = [d, videoSource](time_t t, qint64 i){
+    auto distributor = d->videoSources.value(videoSource);
+    Q_ASSERT(distributor);
+
+    distributor->requestFrame(videoSource, t, SeekExact, i);
+  };
+
+  auto* const newDistributor = new VideoDistributor{this};
+
+  auto const first = d->videoSources.isEmpty();
+  d->videoSources.insert(videoSource, newDistributor);
+  connect(this, &VideoController::timeSelected,
+          videoSource, fetch);
+
+  d->timesDirty = true;
+  if (!first && d->timeIsValid)
+  {
+    fetch(d->time, -1);
   }
+
+  videoSource->start();
+
+  emit this->videoSourcesChanged();
+
+  return newDistributor;
 }
 
 // ----------------------------------------------------------------------------
 void VideoController::removeVideoSource(VideoSource* videoSource)
 {
   QTE_D();
-  if (d->videoSources.remove(videoSource))
+
+  if (auto* const distributor = d->videoSources.take(videoSource))
   {
+    disconnect(videoSource, nullptr, this, nullptr);
     disconnect(this, nullptr, videoSource, nullptr);
+    delete distributor;
```

Comment:
Ugh... is there a reason why we can't use a smart pointer? I hate having code that manually `delete`s anything...

Reply:
It... seems plausible... PTAL at ebce51ed373dd0b8a75e0860b651fe838e62f6fe and let me know if you think it's worth it.